EIPS/eip-7412.md

```diff
@@ -0,0 +1,167 @@
+discussions-to: https://ethereum-magicians.org/t/erc-7412-on-demand-off-chain-data-retrieval/15346
+status: Draft
+type: Standards Track
+category: ERC
+created: 2023-07-26
+---
+
+## Abstract
+
+Contracts may require off-chain data that is not already available on chain during execution. A smart contract function could implement the standard proposed here by reverting with `error OracleDataRequired(address oracleContract, bytes oracleQuery)`. Clients supporting this standard would recognize this error message during a simulation of the request, query the specified decentralized oracle network for signed data, and instead stage a transaction with a multicall that prepends the verification of the required off-chain data. The data would be written on-chain during verification to a smart contract for the subsequent call to read, avoiding the error.
+
+## Motivation
+
+Ethereum's scaling roadmap involves a series of separate execution contexts for smart contract code (including layer two and layer three scaling solutions). This makes the ability to read data across multiple chains crucial to the construction of scalable applications. Also, for decentralized finance protocols that rely on price data, it is not reasonable to expect oracle networks will be able to continuously push fresh data to every layer two and layer three network for an arbitrary number of price feeds.
+
+Cross-chain bridges are being developed, adhering to standards such as [ERC-6170](./eip-6170.md), where smart contract functions can write data to other chains. There is a need for a similar standard that enables reading data from other chains. This standard can be generalized for reading any off-chain data from a decentralized oracle network, including price feeds.
+
+With standards for both writing and reading cross-chain data, protocol developers will be able to create abstractions for asynchronicity (a topic thoroughly explored in other software engineering contexts). This will enable the development of highly sophisticated protocols that do not suffer from scaling constraints.
+
+[ERC-3668](./eip-3668.md) introduced the use of reverts for requiring off-chain data, but there are various challenges introduced by the specifics of that standard which are outlined in the _Rationale_ section below. By leveraging multicalls rather than callback functions, the standard proposed here is able to overcome some of these constraints.
+
+## Specification
+
+A contract implementing this standard MUST revert with the following error whenever off-chain data is required:
+
+```solidity
+error OracleDataRequired(address oracleContract, bytes oracleQuery)
+```
+
+`oracleQuery` specifies the off-chain data that is being required. Valid data formats for this parameter are specific to the oracle ID used. This might include chain id, contract address, function signature, payload, and timestamp/"latest" for cross-chain reads. For price feeds, it could include a ticker symbol and timestamp/"latest".
```

Comment:
"oracle ID" hasn't been defined yet, so maybe this is more clear:

```suggestion
`oracleQuery` specifies the off-chain data that is being required. Valid data formats for this parameter are specific to the oracle contract used. This might include chain id, contract address, function signature, payload, and timestamp/"latest" for cross-chain reads. For price feeds, it could include a ticker symbol and timestamp/"latest".
```

I could also be entirely incorrect here :rofl:

Reply:
Ah, I was entirely incorrect. I'd _maybe_ suggest moving the definition of oracle ID up a bit so that readers don't make the same mistake I did.